pkg/nsx/services/expand.go

```diff
@@ -0,0 +1,310 @@
+) {
+	var addr []util.PortAddress
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			log.V(2).Info("resolvePodPort", "namespace", pod.Namespace, "pod_name", pod.Name,
+				"port_name", port.Name, "containerPort", port.ContainerPort,
+				"protocol", port.Protocol, "podIP", pod.Status.PodIP)
+			if port.Name == spPort.Port.String() && port.Protocol == spPort.Protocol {
+				if pod.Status.Phase != "Running" {
+					errMsg := fmt.Sprintf("pod %s/%s is not running", pod.Namespace, pod.Name)
+					return nil, util2.PodNotRunning{Desc: errMsg}
+				}
+				if pod.Status.PodIP == "" {
+					errMsg := fmt.Sprintf("pod %s/%s ip not initialized", pod.Namespace, pod.Name)
+					return nil, util2.PodIPNotFound{Desc: errMsg}
+				}
+				addr = append(
+					addr,
+					util.PortAddress{Port: int(port.ContainerPort), IPs: []string{pod.Status.PodIP}},
+				)
+			}
+		}
+	}
+	return addr, nil
+}
+
+// Build an ip set group for NSX.
+func (service *SecurityPolicyService) buildRuleIPSetGroup(
+	obj *model.Rule,
+	ips []string,
+) *model.Group {
+	ipSetGroup := model.Group{}
+
+	ipSetGroupID := fmt.Sprintf("%s_ipset", *obj.Id)
+	ipSetGroup.Id = &ipSetGroupID
+	ipSetGroupName := fmt.Sprintf("%s-ipset", *obj.DisplayName)
+	ipSetGroup.DisplayName = &ipSetGroupName
+
+	addresses := data.NewListValue()
+	for _, ip := range ips {
+		addresses.Add(data.NewStringValue(ip))
+	}
+
+	blockExpression := data.NewStructValue(
+		"",
+		map[string]data.DataValue{
+			"resource_type": data.NewStringValue("IPAddressExpression"),
+			"ip_addresses":  addresses,
+		},
+	)
+	ipSetGroup.Expression = append(ipSetGroup.Expression, blockExpression)
+	return &ipSetGroup
+}
+
+// Different direction rule decides different target of the traffic, we should carefully get
+// the destination pod selector and namespaces. Named port only cares about the destination
+// pod selector, so we use policy's AppliedTo or rule's AppliedTo when "IN" direction and
+// rule's DestinationSelector when "OUT" direction.
+func (service *SecurityPolicyService) getPodSelector(obj *v1alpha1.SecurityPolicy,
+	rule *v1alpha1.SecurityPolicyRule,
+) (*client.ListOptions, []string, error) {
+	// Port means the target of traffic, so we should select the pod by rule direction,
+	// as for IN direction, we judge by rule's or policy's AppliedTo,
+	// as for OUT direction, then by rule's destinations.
+	// LabelSelect may return multiple namespaces
+	var namespaces []string
+	selector := client.ListOptions{}
+	var labelSelector labels.Selector
+	ruleDirection, err := getRuleDirection(rule)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if ruleDirection == "IN" {
+		if len(obj.Spec.AppliedTo) > 0 {
+			for _, target := range obj.Spec.AppliedTo {
+				if target.PodSelector != nil {
+					labelSelector, err = meta1.LabelSelectorAsSelector(target.PodSelector)
+					if err != nil {
+						return nil, nil, err
+					}
+				}
+			}
+		} else if len(rule.AppliedTo) > 0 {
```

Comment:
Is it possible `rule.AppliedTo` and `obj.Spec.AppliedTo` both have items?

Reply:
obj level is preferred to rule level

Reply:
Then you need to check rule.AppliedTo first?

Reply:
I mean, obj level is first, so the order is right.

Reply:
Doc says.

**appliedTo**: is a list of policy targets to apply rules. As the CRD is namespaced
scope, `vmSelector` or `podSelector` will be selected from the Namespace where the
CR is created. `vmSelector` and `podSelector` cannot be in one entry as it would
not select any workload. We can also have `appliedTo` in each rule entry, but if
there is policy level `appliedTo`, it will take precedence over rule level.